cartographer_rviz/cartographer_rviz/drawable_trajectory.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2016 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_RVIZ_SRC_DRAWABLE_TRAJECTORY_H_
+#define CARTOGRAPHER_RVIZ_SRC_DRAWABLE_TRAJECTORY_H_
+
+#include <future>
+
+#include "rviz/display_context.h"
+#include "rviz/frame_manager.h"
+#include "rviz/properties/bool_property.h"
+
+#include "cartographer/mapping/id.h"
+#include "cartographer_rviz/drawable_submap.h"
+
+namespace cartographer_rviz {
+class DrawableTrajectory : public QObject {
+  Q_OBJECT
+
+ public:
+  DrawableTrajectory(const int trajectory_id, ::rviz::Property* submap_category,
+                     Ogre::SceneManager* scene_manager, const bool visible);
+  ~DrawableTrajectory() override;
+  DrawableTrajectory(const DrawableTrajectory&) = delete;
+  DrawableTrajectory& operator=(const DrawableTrajectory&) = delete;
+
+  bool visibility() const { return property_->getBool(); }
+  void set_visibility(const bool visibility) { property_->setBool(visibility); }
+
+  const std::map<int, std::unique_ptr<DrawableSubmap>>& submaps() const {
```

Comment:
This is inconsistent - you are handing out the submaps() to check various conditions, i.e. how many requests are in-flight, did Cartographer reset, did the trajectory vanish. But you also hand out a mutable version of your vector that SubmapsList deletes stuff out of.

On the other site you have `UpdateSubmap` that just calls through to the submap.

I see two possible design directions here:

1) Either make this class really own the submaps - never hand them out. Add `NumOngoingRequests`, `SetAlpha`, `processMessage` which should get rid of `AddSubmap` and `UpdateSubmap`. Change SubmapsDisplay to delegate the work to this class.
2) Alternatively (and probably easier and shorter): change SubmapsDisplay to contain a private `struct Trajectory` that is no longer just a `std::pair` and holds the boolean property you want here as well. Then handle all the logic in SubmapsDisplay.